File: src/main/java/com/github/highright1234/glacier/protocol/packet/play/server/EncryptionRequestPacket.java
package com.github.highright1234.glacier.protocol.packet.play.server;

import com.github.highright1234.glacier.protocol.MinecraftPacket;
import io.netty.buffer.ByteBuf;
import lombok.Data;
import lombok.EqualsAndHashCode;

@EqualsAndHashCode(callSuper = true)
@Data
public class EncryptionRequestPacket extends MinecraftPacket {

    private String serverID = "";
    private byte[] publicKey = new byte[0];
    private byte[] verifyToken = new byte[0];

    @Override
    public void write(ByteBuf buf) {
        writeString(serverID, buf);
        writeVarInt(publicKey.length, buf);
        buf.writeBytes(publicKey);
        writeVarInt(verifyToken.length, buf);
        buf.writeBytes(verifyToken);
    }

    @Override
    public void read(ByteBuf buf) {
        serverID = readString(buf);
        int publicKeyLength = readVarInt(buf);
        publicKey = buf.readBytes(publicKeyLength).array();
        int verifyTokenLength = readVarInt(buf);
        verifyToken = buf.readBytes(verifyTokenLength).array();
    }
}

File: src/main/kotlin/com/github/highright1234/glacier/protocol/packet/login/server/EncryptionRequestPacket.kt
package com.github.highright1234.glacier.protocol.packet.login.server

import com.github.highright1234.glacier.protocol.MinecraftPacket
import kotlin.Throws
import java.lang.Exception
import io.netty.buffer.ByteBuf

data class EncryptionRequestPacket(
    var serverID: String = "",
    var publicKey : ByteArray = ByteArray(0),
    var verifyToken : ByteArray = ByteArray(0)
) : MinecraftPacket() {


    override fun write(buf: ByteBuf) {
        writeString(serverID, buf)
        writeVarInt(publicKey.size, buf)
        buf.writeBytes(publicKey)
        writeVarInt(verifyToken.size, buf)
        buf.writeBytes(verifyToken)
    }

    @Throws(Exception::class)
    override fun read(buf: ByteBuf) {
        serverID = readString(buf)
        val publicKeyLength: Int = readVarInt(buf)
        publicKey = buf.readBytes(publicKeyLength).array()
        val verifyTokenLength: Int = readVarInt(buf)
        verifyToken = buf.readBytes(verifyTokenLength).array()
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as EncryptionRequestPacket

        if (serverID != other.serverID) return false
        if (!publicKey.contentEquals(other.publicKey)) return false
        if (!verifyToken.contentEquals(other.verifyToken)) return false

        return true
    }

    override fun hashCode(): Int {
        var result = serverID.hashCode()
        result = 31 * result + publicKey.contentHashCode()
        result = 31 * result + verifyToken.contentHashCode()
        return result
    }
}